module/zfs/zfs_vnops.c

```diff
@@ -941,7 +941,7 @@ zfs_iput_async(struct inode *ip)
 	ASSERT(os != NULL);
 
 	if (atomic_read(&ip->i_count) == 1)
-		taskq_dispatch(dsl_pool_iput_taskq(dmu_objset_pool(os)),
+		(void) taskq_dispatch(dsl_pool_iput_taskq(dmu_objset_pool(os)),
```

Comment:
This is also a real potential issue.  In the case where the dispatch fails we don't have any safe option except retrying the dispatch and hoping for the best.  We can't call `iput()` synchronously because it's possible it will re-enter the vfs and result in a deadlock.  This is why we have an async version in the first place.  And we can't ignore the error because we'll leak a reference and no longer be able to unmount the filesystem.

The good news is that this should be completely impossible given the current taskq_dispatch implementation and flags passed.  I'd suggest the dropping this hunk from this PR and opening a new one with this alternate untested change.


```diff
diff --git a/module/zfs/zfs_vnops.c b/module/zfs/zfs_vnops.c
index 887b7e1..9ac1a3f 100644
--- a/module/zfs/zfs_vnops.c
+++ b/module/zfs/zfs_vnops.c
@@ -932,19 +932,31 @@ zfs_write(struct inode *ip, uio_t *uio, int ioflag, cred_t
 }
 EXPORT_SYMBOL(zfs_write);
 
+/*
+ * Drop a reference on the passed inode asynchronously.  This ensures
+ * that the caller will never drop the last reference on an inode in
+ * the current context.  Doing so while holding open a tx could result
+ * in a deadlock if iput_final() re-enters the filesystem code.
+ */
 void
 zfs_iput_async(struct inode *ip)
 {
        objset_t *os = ITOZSB(ip)->z_os;
+       taskqid_t id = 0;
+
 
        ASSERT(atomic_read(&ip->i_count) > 0);
        ASSERT(os != NULL);
 
-       if (atomic_read(&ip->i_count) == 1)
-               taskq_dispatch(dsl_pool_iput_taskq(dmu_objset_pool(os)),
-                   (task_func_t *)iput, ip, TQ_SLEEP);
-       else
+       if (atomic_read(&ip->i_count) == 1) {
+               while (id == 0) {
+                       id = taskq_dispatch(
+                           dsl_pool_iput_taskq(dmu_objset_pool(os)),
+                           (task_func_t *)iput, ip, TQ_SLEEP);
+               }
+       } else {
                iput(ip);
+       }
 }
 
 void
```